TripalCultivate/TripalCultivate-Phenotypes
# G4.100 split row into columns

## Issue #100 

## Motivation



Create a method that will split or explode a line from the data file into columns using a delimiter chosen based on the mime type of the file.

This PR will 
1. update the importer form validate to pass a string value (the raw data row/line) to validateRawRow() in the importer
2. provide a static splitRowIntoColumns() method so the importer can split the row based on the mime type of the file.
3. ensure all automated testing checks this new approach.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

ValidatorBaseTest - testSplitRowIntoColumns()
1. Test case where no delimiter
2. The method return the same line (delimiter could not split the row)
3. Test that the line was split and that each value was sanitized
4. Test basic row input
5. Do validation 1-4 for all supported mime types.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

In my site/devel/php of a working T4 clone, paste the code block and execute. The line array and/or the delimiter character can be altered to other value to test other set of values.

```
// LINE:
$line = ['Value A', 'Value "B"', 'Value \'C\'', 'Value D ', ' Value E', ' Value F ', ' Value G           '];

// MIME TYPE:
$mime_type = 'text/tab-separated-values';
$delimiter = "\t";

$str_line = implode($delimiter, $line);
$split_row = \Drupal\trpcultivate_phenotypes\TripalCultivateValidator\TripalCultivatePhenotypesValidatorBase::splitRowIntoColumns($str_line, $mime_type);

print_r($split_row);
```

The method includes sanitizing value from quotes (double and single) as well as leading and trailing spaces.

The code should create the array below where the extracted values are sanitized and in an array.

```

Array
(
    [0] => Value A
    [1] => Value B
    [2] => Value C
    [3] => Value D
    [4] => Value E
    [5] => Value F
    [6] => Value G
)

```
